src/hotspot/share/oops/compressedKlass.hpp

```diff
@@ -56,6 +56,24 @@ class CompressedKlassPointers : public AllStatic {
   //  could use this info to optimize encoding.
   static size_t _range;
 
+  // Reduce the number of loads generated to decode an nKlass by packing all
+  // relevant information (flag, base, shift) into a single 64-bit word. The
+  // encoding base is aligned to metaspace reserve alignment (16Mb), so enough
+  // space to hide additional info away in low bytes. It even allows us the luxury
+  // to use 8 bits for shift and flag each, resulting in 8-bit moves used by the
+  // compiler.
+  // - Bit  [0-7]   shift
+  // - Bit  8       UseCompressedOops
+  // - Bits [16-64] the base.
+  static uint64_t _combo;
+  static constexpr int base_alignment = 16;
+  static constexpr uint64_t mask_base = ~right_n_bits(base_alignment);
+  static constexpr int shift_bitlen = 8; // read with a mov8
+  static constexpr int bitpos_useccp = shift_bitlen;
+  static address  base_from_combo()  { return  (address)(_combo & mask_base); }
+  static int      shift_from_combo() { return  (int)(_combo & right_n_bits(shift_bitlen)); }
```

Comment:
```suggestion
  static int      shift() { return  (int)(_combo & right_n_bits(shift_bitlen)); }
```

What's the point of the `_from_combo` functions?

Reply:
They are only used when asserting correctness after initialization. But you are right, I could use `base()` directly.